### ISSUE TITLE
Package.json being overwritten by CLI tool

### ISSUE THREAD

**Issue by thomasjohnkane** [laravel/nova-issues#299]
Hello,

Nova v1.0.6

When generating a new card via CLI, it sometimes overwrites the package.json file in the root of my project:

<img width="936" alt="screen shot 2018-08-28 at 4 29 42 pm" src="https://user-images.githubusercontent.com/8945177/44752089-993ada00-aadf-11e8-9e76-d4302b5360bb.png">

This also happened earlier when I created a tool. However, I have created a tool without this happening. Am I going crazy?

`php artisan nova:card thomasjohnkane/create-email-card` is all I ran. 

Thanks,

**Comment by davidhemphill** [comment]
This is what happens when a tool is generated.

**Comment by ghaskell** [comment]
I think you misunderstand @davidhemphill, it's not overwriting the package.json inside nova-components, it's overwriting the package.json for the application.  The key feature of nova is that it does not impact or destroy the core application but that's exactly what this does.    Please repopen.

**Comment by davidhemphill** [comment]
I'm not misunderstanding anything. That's the way we designed it.

**Comment by ghaskell** [comment]
I'm not able to reproduce consistently but will open a new issue when I manage to figure out the combination of circumstances that causes this to occur.  

Just so we're clear, the Nova tool is **removing** all the dependencies from my application when generating tools.  It's not appending, it's overwriting.

My initial package.json looks like this:

```
{
    "private": true,
    "scripts": {
        "dev": "cross-env NODE_ENV=development node_modules/webpack/bin/webpack.js --progress --hide-modules --config=node_modules/laravel-mix/setup/webpack.config.js",
        "watch": "cross-env NODE_ENV=development node_modules/webpack/bin/webpack.js --watch --progress --hide-modules --config=node_modules/laravel-mix/setup/webpack.config.js",
        "hot": "cross-env NODE_ENV=development node_modules/webpack-dev-server/bin/webpack-dev-server.js --inline --hot --config=node_modules/laravel-mix/setup/webpack.config.js",
        "production": "cross-env NODE_ENV=production node_modules/webpack/bin/webpack.js --progress --hide-modules --config=node_modules/laravel-mix/setup/webpack.config.js"
    },
    "dependencies": {
        "axios": "^0.15.2",
        "bootstrap": "^4.0",
        "chart.js": "^2.7.2",
        "chartist": "^0.11.0",
        "cross-env": "^3.2.3",
        "jquery": "^2.1.4",
        "js-cookie": "^2.1.0",
        "laravel-mix": "1.*",
        "leaflet": "^1.3.3",
        "leaflet-curve": "^1.0.0",
        "lodash": "^4.17.4",
        "moment": "^2.10.6",
        "moment-countdown": "^0.0.3",
        "popper.js": "^1.12",
        "promise": "^7.1.1",
        "rtlcss": "^2.2",
        "smoothscroll-polyfill": "^0.4.3",
        "sweetalert": "^1.1.3",
        "tippy.js": "^2.5.3",
        "urijs": "^1.17.0",
        "v-autocomplete": "^1.8.2",
        "vue": "2.*",
        "vue-select": "^2.4.0",
        "vue-typeahead": "^2.3.2",
        "vuex": "^3.0.1"
    },
    "devDependencies": {
        "@fortawesome/fontawesome-pro": "^5.1.0-4"
    }
}
```

When I create a tool, the desired state would be this:
```
{
    "private": true,
    "scripts": {
        "dev": "cross-env NODE_ENV=development node_modules/webpack/bin/webpack.js --progress --hide-modules --config=node_modules/laravel-mix/setup/webpack.config.js",
        "watch": "cross-env NODE_ENV=development node_modules/webpack/bin/webpack.js --watch --progress --hide-modules --config=node_modules/laravel-mix/setup/webpack.config.js",
        "hot": "cross-env NODE_ENV=development node_modules/webpack-dev-server/bin/webpack-dev-server.js --inline --hot --config=node_modules/laravel-mix/setup/webpack.config.js",
        "production": "cross-env NODE_ENV=production node_modules/webpack/bin/webpack.js --progress --hide-modules --config=node_modules/laravel-mix/setup/webpack.config.js",
        **"build-image-picker": "cd nova-components/ImagePicker && npm run dev",**
        **"build-image-picker-prod": "cd nova-components/ImagePicker && npm run prod"**
    },
    "dependencies": {
        "axios": "^0.15.2",
        "bootstrap": "^4.0",
        "chart.js": "^2.7.2",
        "chartist": "^0.11.0",
        "cross-env": "^3.2.3",
        "jquery": "^2.1.4",
        "js-cookie": "^2.1.0",
        "laravel-mix": "1.*",
        "leaflet": "^1.3.3",
        "leaflet-curve": "^1.0.0",
        "lodash": "^4.17.4",
        "moment": "^2.10.6",
        "moment-countdown": "^0.0.3",
        "popper.js": "^1.12",
        "promise": "^7.1.1",
        "rtlcss": "^2.2",
        "smoothscroll-polyfill": "^0.4.3",
        "sweetalert": "^1.1.3",
        "tippy.js": "^2.5.3",
        "urijs": "^1.17.0",
        "v-autocomplete": "^1.8.2",
        "vue": "2.*",
        "vue-select": "^2.4.0",
        "vue-typeahead": "^2.3.2",
        "vuex": "^3.0.1"
    },
    "devDependencies": {
        "@fortawesome/fontawesome-pro": "^5.1.0-4"
    }
}
```

And that works 90% of the time.  But sometimes, the tool is deleting the entire package.json file and the only lines that remain are the two additions:
```
    "scripts": {
        "build-image-picker": "cd nova-components/ImagePicker && npm run dev",
        "build-image-picker-prod": "cd nova-components/ImagePicker && npm run prod"
    }
```
That's the whole file.  The CLI tool will delete all the entire dependency object, and all the build scripts except for the ones that it is injecting.  

Don't mean to rant at you like a crazy person, I know y'all are buried in issues but I'm 100% sure this is not the behavior you intended.

**Comment by davidhemphill** [comment]
Ok, I see now what y'all are saying. Sorry for being so short. 😬 And yeah that doesn't seem like desired behavior. When this happens, can you see if that command does it repeatedly?

**Comment by ghaskell** [comment]
It does not.  There is some set of circumstances that causes that behavior.  I've had it happen twice and then attempted for about an hour to recreate.  It seems to be some combination of not installing dependencies in the CLI but then running composer.  

I haven't had a chance to dig through src to see what the cause might be but if I have a chance I'll do so.

**Comment by thomasjohnkane** [comment]
I think I may have this sorted.

If there is invalid json in your root `package.json` file for whatever reason (maybe a missing or extra comma, etc), then the first line of the following function returns `NULL` instead of the expected array.

```
/**
     * Add a path repository for the tool to the application's composer.json file.
     *
     * @return void
     */
    protected function addScriptsToNpmPackage()
    {
        $package = json_decode(file_get_contents(base_path('package.json')), true);

        $package['scripts']['build-'.$this->toolName()] = 'cd '.$this->relativeToolPath().' && npm run dev';
        $package['scripts']['build-'.$this->toolName().'-prod'] = 'cd '.$this->relativeToolPath().' && npm run prod';

        file_put_contents(
            base_path('package.json'),
            json_encode($package, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES)
        );
    }
```

As a result, the next lines just append the new scripts to nothing and rewrite the package.json file to have the reported result.

I'm not sure if that's what caused the issue for @ghaskell, but it seems to cause the result repeatedly. If it is the only cause, I'm not sure we could really call this an issue with nova. Perhaps a simple NULL check to confirm would be nice though?

```
/**
     * Add a path repository for the tool to the application's composer.json file.
     *
     * @return void
     */
    protected function addScriptsToNpmPackage()
    {
        $package = json_decode(file_get_contents(base_path('package.json')), true);
        
        if ($package OR $this->confirm("Package.json appears to be empty or invalid. Would you like to overrite it?", true)) {
            $package['scripts']['build-'.$this->toolName()] = 'cd '.$this->relativeToolPath().' && npm run dev';
            $package['scripts']['build-'.$this->toolName().'-prod'] = 'cd '.$this->relativeToolPath().' && npm run prod';

            file_put_contents(
                base_path('package.json'),
                json_encode($package, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES)
            );
        }
    }
```

**Comment by ghaskell** [comment]
I was working on fixing another bug when it occurred so it's absolutely possible that I had an errant comma which I then repaired before the subsequent run.

When the CLI tool runs, it asks whether you want to install your dependencies and if you were to say yes, yarn or npm would spit out an invalid json error but when this occurred for me I did hit no for this option. 

So it seems that steps to reproduce would be: 

1. Have invalid json in your package.json.
2. Install a tool or resource-tool
3. During installation, decline to run yarn/npm install.
4. Run composer update.

Admittedly, this does seem pretty edge-casey to me as well (and a direct result of an error on my end) but a null check or a `json_last_error()` check that throws an exception as @thomasjohnkane suggests seems like a reasonable safeguard.
